### PR TITLE
Reference for `conan.tools.meson` classes and functions

### PR DESCRIPTION
**Pull request by franramirez688**
Depends on: https://github.com/conan-io/conan/pull/11237

### CODE REVIEW

File: reference/tools/meson/mesontoolchain.rst [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+* ``ar == $TOOLCHAIN/bin/$TARGET$API-clang++``
+
+Where:
+
+* ``$TOOLCHAIN``: ``[NDK_PATH]/toolchains/llvm/prebuilt/[OS_BUILD]-x86_64/bin``.
+* ``$TARGET``: target triple, e.g., for ``armv8`` will be ``aarch64-linux-android``.
+* ``$API``: Android API version.
+
+Besides that, you'll always be able to change any of these variables before being applied thanks
+to the ``MesonToolchain`` class interface. For instance:
```
[the remark this answers is not in the document]
Comment (lasote) [reply reference/tools/meson/mesontoolchain.rst:184]: Which variables? only the ones for cross-building or all of the documented in the class below? Not sure if it is clear enough.

**Review by lasote — CHANGES_REQUESTED** [review]

File: reference/tools/meson/mesontoolchain.rst [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+.. code:: bash
+
+    # Lets start in the folder containing the conanfile.py
+    $ mkdir build && cd build
+    # Install both debug and release deps and create the toolchain
+    $ conan install ..
+    # the build type Release is encoded in the toolchain already.
+    # This conan_meson_native.iniis specific for release
+    $ meson setup --native-file conan_meson_native.ini build .
+    $ meson compile -C build
```
[the remark this answers is not in the document]
Comment (memsharded) [reply reference/tools/meson/mesontoolchain.rst:155]: Agree

File: examples/tools/meson/mesontoolchain/developer_flow.rst [hunk 3]
```diff
@@ -0,0 +1,20 @@
+.. _examples_tools_meson_mesontoolchain_developer_flow:
+
+MesonToolchain: using in developer flow
+----------------------------------------
+
+One of the advantages of using Conan toolchains is that they can help to achieve the exact same build
```
Comment (lasote) [inline examples/tools/meson/mesontoolchain/developer_flow.rst:6]: The idea of the examples is having the "real" one in the https://github.com/conan-io/examples2 repository. right @czoido ?
This example makes no much sense without seeing the recipe + build files + project...

Comment (franramirez688) [reply examples/tools/meson/mesontoolchain/developer_flow.rst:6]: Oh, ok. Then, I'll remove it for now till I have added a good example there 😉

Comment (lasote) [reply examples/tools/meson/mesontoolchain/developer_flow.rst:6]: yeah, I did the same. It is kind of tough to complete the circle.

File: reference/tools/meson/mesontoolchain.rst [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+conan_meson_native.ini
++++++++++++++++++++++++++
+
+This file contains the definitions of all the Meson properties related to the Conan options
+and settings for the current package, platform, etc. This includes but is not limited to the following:
+
+* Detection of ``default_library`` from Conan settings.
+
+    * Based on existance/value of a option named ``shared``.
```
Comment (memsharded) [inline reference/tools/meson/mesontoolchain.rst:51]:
```suggestion
    * Based on existence/value of an option named ``shared``.
```

File: reference/tools/meson/mesontoolchain.rst [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+.. code:: python
+
+    def generate(self):
+        tc = MesonToolchain(self)
+        tc.definitions["MYVAR"] = "MyValue"
+        tc.generate()
+
+This is translated to:
+
+- One project options definition for ``MYVAR`` in ``conan_meson_native.init`` or ``conan_meson_cross.ini`` file.
```
Comment (memsharded) [inline reference/tools/meson/mesontoolchain.rst:94]:
```suggestion
- One project options definition for ``MYVAR`` in ``conan_meson_native.ini`` or ``conan_meson_cross.ini`` file.
```

File: reference/tools/meson/mesontoolchain.rst [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    def generate(self):
+        tc = MesonToolchain(self)
+        tc.preprocessor_definitions["MYDEF"] = "MyValue"
+        tc.generate()
+
+This is translated to:
+
+- One preprocessor definition for ``MYDEF`` in ``conan_meson_native.ini`` or ``conan_meson_cross.ini`` file.
+
+.. note::
```
[the remark this answers is not in the document]
Comment (franramirez688) [reply reference/tools/meson/mesontoolchain.rst:112]: My fault. Bad copy&paste from Conan 1.x

File: reference/tools/meson/mesontoolchain.rst [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+**Apple**
+
+It adds link flags ``-arch XXX``, ``-isysroot [SDK_PATH]`` and the minimum deployment target flag, e.g., ``-mios-version-min=8.0``
+to the ``MesonToolchain`` ``c_args``, ``c_link_args``, ``cpp_args`` and ``cpp_link_args`` attributes.
+
+**Android**
+
+It initializes the ``MesonToolchain`` ``c``, ``cpp``, and ``ar`` attributes, which are needed to cross-compile for Android. For instance:
+
+* ``c == $TOOLCHAIN/bin/llvm-ar``
```
[the remark this answers is not in the document]
Comment (franramirez688) [reply reference/tools/meson/mesontoolchain.rst:149]: ok, let me rephrase it a little bit

**Review by memsharded — APPROVED** [review]

File: reference/tools/meson/meson.rst [hunk 8]
```diff
@@ -0,0 +1,32 @@
+.. _conan_tools_meson_meson:
+
+Meson
+======
+
+The ``Meson()`` build helper is intended to be used in the ``build()`` and ``package()`` methods, to call Meson commands automatically
+when a package is being built directly by Conan (create, install).
```
Comment (czoido) [inline reference/tools/meson/meson.rst:7]:
```suggestion
The ``Meson()`` build helper is intended to be used in the ``build()`` and ``package()`` methods, to call Meson commands automatically.
```
I would stop the sentence there or add other commands that can call the build or package methods like `build` or `export-pkg`, but I don't think that's necessary.

File: reference/tools/meson/mesontoolchain.rst [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        def generate(self):
+            tc = MesonToolchain(self)
+            tc.preprocessor_definitions["MYDEFINE"] = "MYDEF_VALUE"
+            tc.generate()
+
+
+.. important::
+
+    The ``MesonToolchain`` only works with the ``PkgConfigDeps`` generator.
+    Please, do not use other generators, as they can have overlapping definitions that can conflict.
```
[the remark this answers is not in the document]
Comment (franramirez688) [reply reference/tools/meson/mesontoolchain.rst:30]: I can rephrase it, but I think it's important to know. The examples are fine to know how it works, but I guess that you should point it out here as well.

[the remark this answers is not in the document]
Comment (lasote) [reply reference/tools/meson/mesontoolchain.rst:30]: Yes, I think a rephrase would be ok. Something saying, "if you have dependencies" is the missing piece.

**Review by lasote — APPROVED** [review]

File: reference/tools/meson/mesontoolchain.rst [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+Cross-building for Apple and Android
+-------------------------------------
+
+The ``MesonToolchain`` adds all the flags required to cross-compile for Apple (MacOS M1, iOS, etc.) and Android.
+
+**Apple**
+
+It adds link flags ``-arch XXX``, ``-isysroot [SDK_PATH]`` and the minimum deployment target flag, e.g., ``-mios-version-min=8.0``
+to the ``MesonToolchain`` ``c_args``, ``c_link_args``, ``cpp_args``, and ``cpp_link_args`` attributes, given the
+Conan settings for any Apple OS (iOS, watchOS, etc.) and the ``tools.apple:sdk_path`` configuration value like it's showed
```
Comment (lasote) [inline reference/tools/meson/mesontoolchain.rst:139]:
```suggestion
Conan settings for any Apple OS (iOS, watchOS, etc.) and the ``tools.apple:sdk_path`` configuration value like it's shown
```

File: reference/tools/meson/mesontoolchain.rst [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    compiler.libcxx = libc++
+
+    [conf]
+    tools.apple:sdk_path=/my/path/to/iPhoneOS.sdk
+
+
+**Android**
+
+It initializes the ``MesonToolchain`` ``c``, ``cpp``, and ``ar`` attributes, which are needed to cross-compile for Android, given the
+Conan settings for Android and the ``tools.android:ndk_path`` configuration value like it's showed
```
Comment (lasote) [inline reference/tools/meson/mesontoolchain.rst:161]:
```suggestion
Conan settings for Android and the ``tools.android:ndk_path`` configuration value like it's shown
```

**Review by lasote — APPROVED** [review]

File: reference/tools/meson/mesontoolchain.rst [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    The ``MesonToolchain`` only works with the ``PkgConfigDeps`` generator.
+    Please, do not use other generators, as they can have overlapping definitions that can conflict.
+
+
+Generated files
+-----------------
+
+This generates the following files after a :command:`conan install` (or when building the package
+in the cache) with the information provided in the ``generate()`` method as well as information
+translated from the current ``settings``, ``conf``, etc.:
```
Comment (czoido) [inline reference/tools/meson/mesontoolchain.rst:38]:
```suggestion
The ``MesonToolchain`` generates the following files after a :command:`conan install` (or when building the package
in the cache) with the information provided in the ``generate()`` method as well as information
translated from the current ``settings``, ``conf``, etc.:
```

File: reference/tools/meson/mesontoolchain.rst (hunk 9 above)
Comment (czoido) [inline reference/tools/meson/mesontoolchain.rst:30]:
```suggestion
    When your recipe has dependencies ``MesonToolchain`` only works with the ``PkgConfigDeps`` generator.
    Please, do not use other generators, as they can have overlapping definitions that can conflict.
```

**Review by czoido — APPROVED** [review]
